File: backend/app/controller/asset.go
package controller

import (
	"math"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/phongsakk/finn4u-back/app/database"
	"github.com/phongsakk/finn4u-back/app/database/models"
	"github.com/phongsakk/finn4u-back/app/request"
	"github.com/phongsakk/finn4u-back/types"
	"github.com/phongsakk/finn4u-back/utils"
)

func GetAsset(c *gin.Context) {
	var page = 0
	var take = 20
	var offset = utils.Offset(page, take)
	var response []models.Asset

	db, err := database.Conn()
	if err != nil {
		c.JSON(http.StatusInternalServerError, types.Response{
			Code:  http.StatusInternalServerError,
			Error: utils.NullableString(err.Error()),
		})
		return
	}
	defer database.Close(db)

	// Count all assets before applying pagination
	var totalAssets int64
	if err := db.Model(&models.Asset{}).Count(&totalAssets).Error; err != nil {
		c.JSON(http.StatusInternalServerError, types.Response{
			Code:  http.StatusInternalServerError,
			Error: utils.NullableString(err.Error()),
		})
		return
	}

	// Fetch assets with pagination
	if err := db.Offset(offset).Limit(take).Order("id").Find(&response).Error; err != nil {
		c.JSON(http.StatusInternalServerError, types.Response{
			Code:  http.StatusInternalServerError,
			Error: utils.NullableString(err.Error()),
		})
		return
	}

	total := len(response)
	totalPage := int64(math.Ceil(float64(totalAssets) / float64(len(response))))
	c.JSON(200, types.Response{
		Code:      http.StatusOK,
		Message:   utils.NullableString("Assets retrieved successfully"),
		Data:      response,
		Page:      &page,
		Limit:     &take,
		Total:     &total,
		TotalPage: &totalPage,
	})
}

func CreateAsset(c *gin.Context) {
	var request request.CreateAssetRequest

	if err := c.ShouldBindBodyWithJSON(&request); err != nil {
		c.JSON(http.StatusBadRequest, types.Response{
			Code:  http.StatusBadRequest,
			Error: utils.NullableString(err.Error()),
		})
		return
	}
	if err := request.Validated(); err != nil {
		c.JSON(http.StatusBadRequest, types.Response{
			Code:  http.StatusBadRequest,
			Error: utils.NullableString(err.Error()),
		})
		return
	}

	db, err := database.Conn()
	if err != nil {
		c.JSON(http.StatusInternalServerError, types.Response{
			Code:  http.StatusInternalServerError,
			Error: utils.NullableString(err.Error()),
		})
		return
	}
	defer database.Close(db)
	var asset models.Asset = models.Asset{
		ProvinceID:          request.ProvinceID,
		DistrictID:          request.DistrictID,
		AssetTypeID:         request.AssetTypeID,
		AriaSizeRai:         request.AriaSizeRai,
		AriaSizeNgan:        request.AriaSizeNgan,
		AriaSizeSquareWa:    request.AriaSizeSquareWa,
		Collateral:          request.Collateral,
		ConsignmentPrice:    request.ConsignmentPrice,
		LandTitleDeedNumber: request.LandTitleDeedNumber,
		LandPlotNumber:      request.LandPlotNumber,
		LocationX:           request.LocationX,
		LocationY:           request.LocationY,
		IsMultipleHolder:    request.IsMultipleHolder,
		EndedAt:             request.EndedAt,
		OwnerID:             3,
	}

	if err := db.Where("id =?", request.ProvinceID).First(&models.Province{}).Error; err != nil {
		c.JSON(http.StatusBadRequest, types.Response{
			Code:  http.StatusBadRequest,
			Error: utils.NullableString("Province not found"),
		})
		return
	}

	if err := db.Where("id =?", request.DistrictID).First(&models.District{}).Error; err != nil {
		c.JSON(http.StatusBadRequest, types.Response{
			Code:  http.StatusBadRequest,
			Error: utils.NullableString("District not found"),
		})
		return
	}

	if err := db.Where("id =?", request.AssetTypeID).First(&models.AssetType{}).Error; err != nil {
		c.JSON(http.StatusBadRequest, types.Response{
			Code:  http.StatusBadRequest,
			Error: utils.NullableString("Asset type not found"),
		})
		return
	}

	if err := db.Create(&asset).Error; err != nil {
		c.JSON(http.StatusBadRequest, types.Response{
			Code:  http.StatusBadRequest,
			Error: utils.NullableString(err.Error()),
		})
		return
	}

	c.JSON(201, types.Response{
		Code:    http.StatusCreated,
		Message: utils.NullableString("Asset created successfully"),
		Data:    asset,
	})
}

File: backendv2/src/app/http/controllers/asset.ts
import { CustomHandler, CustomPagingHandler } from "../../../types/http";
import { safeNumber, serializeBigInt } from "../../../utils/data";
import prisma from "../../../utils/prisma";

export const Recommended: CustomPagingHandler = async (req, res) => {
  const page = safeNumber(req.query.page, 1);
  const take = safeNumber(req.query.take, 10);
  const offset = (page - 1) * take;

  const total = await prisma.$queryRawUnsafe<{total_count: number}[]>(`
  SELECT COUNT(*) AS total_count FROM (
    SELECT 1 FROM asset WHERE recommended_at IS NOT NULL
    UNION ALL
    SELECT 1 FROM sell WHERE recommended_at IS NOT NULL
  ) AS combined;
  `);
  const data = await prisma.$queryRawUnsafe<any[]>(`
    select 'ขายฝาก' as sell_type, a.id, at2."name" as asset_type, p."name" as province_name,
    a.gen_id, a.aria_size_rai, a.aria_size_ngan, a.aria_size_square_wa, a.aria_size_square_metre ,
    aa.collatera_price as price, recommended_at
    from asset a
    left join asset_type at2 on at2.id = a.asset_type_id 
    left join province p on p.id = a.province_id
    left join asset_appraisal aa on aa.asset_id = a.id
    where recommended_at is not null
    union
    select st."name" as sell_type ,s.id, at2."name" as asset_type, p."name" as province_name,
    s.gen_id as gen_id, 0 as aria_size_rai, 0 as aria_size_ngan, 0 as aria_size_square_wa, 0 as aria_size_square_metre,
    s.price, recommended_at
    from sell s 
    left join sell_type st on st.id = s.sell_type_id 
    left join asset_type at2 on at2.id = s.asset_type_id 
    left join province p on p.id = s.province_id
    where recommended_at is not null 
    order by recommended_at desc
    offset ${offset} limit ${take}
    `);

    const serial = serializeBigInt(total)
    const total_count = safeNumber(serial[0].total_count)

  res.status(200).json({
    status: true,
    code: 200,
    message: "Recommended assets fetched successfully",
    data: serializeBigInt(data),
    page: page,
    total: data.length,
    total_page: Math.ceil(total_count / take),
    limit: take,
  });
  return;
};
